File: Algorithms/Python/116 Populating Next Right Pointers in Each Node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        def connect_level(nodes):
            if len(nodes) == 0:
                return
            
            sublevel = []
            for i in range(len(nodes) - 1):
                nodes[i].next = nodes[i + 1]
                if nodes[i].left:
                    sublevel.append(nodes[i].left)
                if nodes[i].right:
                    sublevel.append(nodes[i].right)
            
            nodes[-1].next = None
            if nodes[-1].left:
                sublevel.append(nodes[-1].left)
            if nodes[-1].right:
                sublevel.append(nodes[-1].right)
            connect_level(sublevel)
        
        if root is None:
            return None

        nodes = []
        nodes.append(root)
        connect_level(nodes)
        return root


# recursion    
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root is None:
            return root

        left, right = root.left, root.right
        while left and right:
            left.next = right
            left = left.right
            right = right.left
        
        self.connect(root.left)
        self.connect(root.right)
        return root    

File: Algorithms/CPP/116 Populating Next Right Pointers in Each Node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if (root == NULL) {
            return NULL;
        }
        std::vector<Node*> nodes;
        nodes.push_back(root);
        connect(nodes);
        return root;
    }
    void connect(std::vector<Node*> nodes) {
        if (nodes.size() == 0) {
            return;
        }
        std::vector<Node*> subLevel;
        for (int i = 0; i < nodes.size(); ++i) {
            if (i + 1 < nodes.size()) {
                nodes[i]->next = nodes[i + 1];
            }
            if (nodes[i]->left) {
                subLevel.push_back(nodes[i]->left);
            }
            if (nodes[i]->right) {
                subLevel.push_back(nodes[i]->right);
            }
        }
        nodes.back()->next = NULL;
        connect(subLevel);
    }
};


# recursion
class Solution {
public:
    Node* connect(Node* root) {
        if (root == NULL) {
            return NULL;
        }
        Node* left = root->left;
        Node* right = root->right;
        while (left && right) {
            left->next = right;
            left = left->right;
            right = right->left;
        }
        connect(root->left);
        connect(root->right);
        return root;
    }
};
